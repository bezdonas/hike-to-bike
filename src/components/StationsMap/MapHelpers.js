import Leaflet from 'leaflet';

export const initMap = (className, coords, zoom) => {
  // document.head.innerHTML += `<link rel="stylesheet" href="https://unpkg.com/leaflet@1.3.4/dist/leaflet.css"
  //       integrity="sha512-puBpdR0798OZvTTbP4A8Ix/l+A4dHDD0DGqYW6RQ+9jxkRFclaxxQb/SJAWZfWAkuyeQUytO7+7N4QKrDh+drA=="
  //       crossorigin=""/>`;
  const mapInstance = Leaflet.map(className).setView(coords, zoom);

  Leaflet.tileLayer('https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png', {
    attribution:
      '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors',
  }).addTo(mapInstance);

  return mapInstance;
};
