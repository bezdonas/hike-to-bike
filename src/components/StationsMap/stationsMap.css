.stations-map {
  z-index: 90;
  height: 100vh;
}

@media (max-width: 650px) {
  .stations-map {
    height: 75vh;
  }
}
