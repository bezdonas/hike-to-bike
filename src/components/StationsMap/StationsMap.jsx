import { h, Component } from 'preact';
import {
  initMap,
  mapPanTo,
  removeMarker,
  addStationToMap,
  addCurrentPosToMap,
} from '../../MapAdapter/LeafletAdapter.js';
import {
  getStationsPromise,
  parseStations,
  getClosestStations,
} from '../../StationsApi/StationsApi.js';
import './stationsMap.css';

const defaultZoom = 15;
const stationsUpdateIntervalDuration = 10000;

// StationsMapProps = { currentPosition: [lat, lng], showAllStations: boolean }
export default class StationsMap extends Component {
  constructor(props) {
    super(props);

    this.updateStationsInterval;

    this.mapInstance;

    this.currentPositionMarker;
    this.markers = {
      // stationId: LeafletMarker
    };

    this.allStations = [];
    this.closestStations = [];
  }

  addStationsToMap() {
    this.closestStations = getClosestStations(
      this.props.currentPosition,
      this.allStations
    );

    const stationsToShow = this.props.showAllStations
      ? this.allStations
      : this.closestStations;

    stationsToShow.forEach(station => {
      this.markers[station.id] = addStationToMap(this.mapInstance, station);
    });
  }

  removeOldStationsFromMap() {
    const stationsToRemove = this.props.showAllStations
      ? this.allStations
      : this.closestStations;
    stationsToRemove.forEach(station => {
      removeMarker(this.markers[station.id]);
    });
  }

  showCurrentPosition() {
    this.currentPositionMarker = addCurrentPosToMap(
      this.mapInstance,
      this.props.currentPosition
    );
  }

  removeOldCurrentPosition() {
    removeMarker(this.currentPositionMarker);
  }

  getStations() {
    getStationsPromise().then(stations => {
      this.removeOldStationsFromMap();
      this.allStations = stations;
      this.addStationsToMap();
    });
  }

  componentDidMount() {
    this.mapInstance = initMap(
      'stations-map',
      this.props.currentPosition,
      defaultZoom
    );
    this.showCurrentPosition();

    this.getStations();
    this.updateStationsInfoInterval = setInterval(
      this.getStations.bind(this),
      stationsUpdateIntervalDuration
    );
  }

  componentWillUpdate() {
    this.removeOldStationsFromMap();
    this.removeOldCurrentPosition();
  }

  componentDidUpdate() {
    mapPanTo(this.mapInstance, this.props.currentPosition);
    this.showCurrentPosition();
    this.addStationsToMap();
  }

  componentWillUnmount() {
    clearInterval(this.updateStationsInfoInterval);
  }

  render() {
    return <div id="stations-map" className="stations-map" />;
  }
}
