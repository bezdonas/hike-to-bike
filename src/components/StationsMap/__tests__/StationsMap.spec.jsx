import { h, preact } from 'preact';
import 'preact-compat';
import render from 'preact-render-to-json';
import StationsMap from '../StationsMap.jsx';

describe('StationsMap', () => {
  it('renders as expected and matches snapshot with showAllStations', () => {
    const props = {
      showAllStations: true,
      currentPosition: [55, 31],
    };
    expect(render(<StationsMap {...props} />)).toMatchSnapshot();
  });
  it('renders as expected and matches snapshot without showAllStations', () => {
    const props = {
      showAllStations: false,
      currentPosition: [55, 31],
    };
    expect(render(<StationsMap {...props} />)).toMatchSnapshot();
  });
});
