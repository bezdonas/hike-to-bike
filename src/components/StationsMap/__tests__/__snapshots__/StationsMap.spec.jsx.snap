// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StationsMap renders as expected and matches snapshot with showAllStations 1`] = `
<div
  class="stations-map"
  id="stations-map"
/>
`;

exports[`StationsMap renders as expected and matches snapshot without showAllStations 1`] = `
<div
  class="stations-map"
  id="stations-map"
/>
`;
